### ISSUE TITLE
three dot range notation with spaces confuses parser

### ISSUE THREAD

**Issue by 747** [DanielXMoore/Civet#844]
From https://github.com/DanielXMoore/Civet/issues/836#issuecomment-1848893323:

Currently,
```
for [0..1]
  console.log 'Hello Civet!'
for [0 .. 1]
  console.log 'Hello Civet!'
for [0...1]
  console.log 'Hello Civet!'
```
gives
``` ts
for (let i = 0; i <= 1; ++i) {
  console.log("Hello Civet!");
}
for (let i1 = 0; i1 <= 1; ++i1) {
  console.log("Hello Civet!");
}
for (let i2 = 0; i2 < 1; ++i2) {
  console.log("Hello Civet!");
}
```
but
```
for [0 ... 1]
  console.log 'Hello Civet!'
```
returns `ParseError`.

Also,
```
for i in [0 ... list]
  console.log 'Hello Civet!'
```
becomes
``` ts
for (const i in [0(...list)]) {
  console.log("Hello Civet!");
}
```
which seems a pointless function call.

There should be an appropriate parser for `...` used as infix.

**Comment by edemaine** [comment]
Ah, now I see how this is getting parsed: `x ...y` and `x ... y` also have a legitimate parse as `x(...y)`.

I'm afraid this is an inherent ambiguity, so I'd hesitate to force range behavior here.  I can add documentation that there can be no space before `...` in ranges. Open to better ideas though.

**Comment by bbrk24** [comment]
Doesn't CoffeeScript handle this by putting the rest `...` the other way, making the function call `x y...`?

**Comment by edemaine** [comment]
CoffeeScript (2) and Civet [support both](https://coffeescript.org/#try:x%20...y%0Ax%20y...).

CoffeeScript doesn't have typing so it doesn't have a second meaning of `:` to disambiguate from.  Requiring `.=`/`:=` or `let`/`const` in typed declarations is crucial for Civet achieving this disambiguation.

**Comment by bbrk24** [comment]
The alternative is to adopt a different range syntax. Aussie++ is a mostly non-serious language but I do like its ranges: `[0 TO 2]`, `[0 TO 2)`, `(0 TO 2]`, and `(0 TO 2)`. Coming from a mathematical background these are more intuitive than two dots vs three (especially considering that two dots makes `0..toString()` ambiguous), and they technically allow you to describe a fully-open range (though ime that's rarely needed).

**Comment by STRd6** [comment]
We could add a check so that `x ... y` won't parse as function application and require `x ...y` if we wanted to apply the function.

**Comment by STRd6** [comment]
Forbidding `x ... y` as application in all situations seems like the simplest way.

It looks like we already have `a[...x]` as `a.slice(0, x)` while keeping `[... x]` as a spread. This is consistent with CoffeeScript2. https://civet.dev/playground?code=YVsuLi54XQpbLi4ueF0%3D
https://coffeescript.org/#try:a%5B...x%5D%0A%5B...x%5D

**Comment by edemaine** [comment]
Implemented in 0.6.60.
